apps/api/src/domain/models/user-management/user/commands/grant-user-role/grant-user-role.command.integration.spec.ts

```diff
@@ -0,0 +1,184 @@
+                    await assertCommandSuccess(commandAssertionDependencies, {
+                        buildValidCommandFSA: () => ({
+                            type: commandType,
+                            payload: {
+                                userId: existingUser.id,
+                                role,
+                            },
+                        }),
+                        initialState: buildInMemorySnapshot({
+                            users: [
+                                existingUser.clone({
+                                    roles: [getDistinctRole(role)],
+                                }),
+                            ],
+                        }),
+                    });
+                });
+            });
+        });
+    });
+
+    describe('when the command is invalid', () => {
+        describe('when there is no user with the given ID', () => {
+            it('should fail', async () => {
+                await assertCommandError(commandAssertionDependencies, {
+                    buildCommandFSA: () => validCommandFSA,
+                    initialState: buildEmptyInMemorySnapshot(),
+                    checkError: (error: InternalError) => {
+                        expect(error).toBeInstanceOf(CommandExecutionError);
+
+                        expect(error.innerErrors[0]).toEqual(
+                            new AggregateNotFoundError(existingUser.getCompositeIdentifier())
+                        );
+                    },
+                });
+            });
+        });
+
+        describe('when the user already has the given role', () => {
+            it('should return the appropriate error', async () => {
+                await assertCommandError(commandAssertionDependencies, {
+                    buildCommandFSA: () => validCommandFSA,
+                    initialState: buildInMemorySnapshot({
+                        users: [
+                            existingUser.clone({
+                                // the existing user already has the specified role
+                                roles: [validCommandFSA.payload.role],
+                            }),
+                        ],
+                    }),
+                });
+            });
+        });
+
+        describe('when there is a property with an invalid type on the command payload', () => {
+            generateCommandFuzzTestCases(GrantUserRole).forEach(
+                ({ description, propertyName, invalidValue }) => {
+                    describe(`when the property ${propertyName} has the invalid value: ${invalidValue} (${description})`, () => {
+                        it('should fail with the appropriate error', async () => {
+                            await assertCommandFailsDueToTypeError(
+                                commandAssertionDependencies,
+                                propertyName,
+                                invalidValue,
+                                buildInvalidFSA
```

Comment:
I wonder if we could pass the validFSA and new up the InvalidFSAFactory inside `assertCommandFailsDueToTypeError`.